### PR TITLE
students home work. I try the best))

### PR DESCRIPTION
**Pull request by Oleksii1985**

### CODE REVIEW

File: lesson_20_1/students.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    };
+  }
+
+  getFullName() {
+    return `${this.info.name} ${this.info.surname}`;
+  }
+
+  setFullName(fullName) {
+    return fullName.split(" ");
+  }
```
Comment (kosachevlad) [inline lesson_20_1/students.js:18]: wrong implementation of method setFullName. After this method is called, the user's name and surname should change!

File: lesson_20_1/students.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+const alex = new Human({
+  name: "Oleksii",
+  surname: "Voitiuk",
+  age: 37,
+});
+
+console.log(alex);
+console.log(alex.getFullName());
+console.log(alex.setFullName("Oleksii Voitiuk"));
```
Comment (kosachevlad) [inline lesson_20_1/students.js]: You probably didn't understand the setFullName method right away. Let me explain: it's supposed to reassign the user's first and last name to the given instance. You can test it this way :
```javascript
console.log(alex.getFullName()); // 'Oleksii Voitiuk'
alex.setFullName("John Doe");
console.log(alex.name) // 'John'
console.log(alex.surname) // 'Doe'
```

File: lesson_20_1/students.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+class Student extends Human {
+  constructor(info) {
+    super(info);
+
+    this.info = {
+      ...info,
+      marks: info.marks,
+    };
```
Comment (kosachevlad) [inline lesson_20_1/students.js]: marks shouldn't be inside info object.
your constructor should like :
```javascript
constructor(info) {
  super(info);
  this.marks = info.marks;
}
```
and refer to the marks value below the code like this  `this.marks`

File: lesson_20_1/students.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  getAverageMark() {
+    return (
+      this.info.marks.reduce(
```
Comment (kosachevlad) [inline lesson_20_1/students.js]: `this.marks.reduce(`

File: lesson_20_1/students.js [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        (prevValue, currValue) => prevValue + currValue,
+        0
+      ) / this.info.marks.length
```
Comment (kosachevlad) [inline lesson_20_1/students.js]: `this.marks.length`

File: lesson_20_1/students.js [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    );
+  }
+
+  getMinMark() {
+    return Math.min.apply(Math, this.info.marks);
```
Comment (kosachevlad) [inline lesson_20_1/students.js]: 1.  `return Math.min.apply(Math, this.marks);`

2. let me show shortened code with rest operator ``` return Math.min(...this.marks);```

File: lesson_20_1/students.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  getMaxMark() {
+    return Math.max.apply(Math, this.info.marks);
```
Comment (kosachevlad) [inline lesson_20_1/students.js]: same...

File: lesson_20_1/students.js [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+class FakeStudent extends Student {
+  #cheatedMarks = this.#cheat();
+
+  constructor(info) {
+    super(info);
+    this.info = {
+      ...info,
+      marks: this.#cheatedMarks,
```
Comment (kosachevlad) [inline lesson_20_1/students.js]: same mistake as in Student constructor..

File: lesson_20_1/students.js [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+class Teacher extends Human {
+  constructor(info) {
+    super(info);
+
+    this.group = {
+      ...info,
+      studentsList: [vlad, den, marta, sveta, ben, nik],
+    };
+  }
```
Comment (kosachevlad) [inline lesson_20_1/students.js:144]: wrong implementation. Group must be specified in the constructor when creating the teacher instance 
`let John = new Teacher({name: 'John', surname: 'Doe', age: 30, group: [vlad, den, marta, sveta, ben, nik]})`
and your constructor should look like :
```javascript
constructor(info) {
    super(info);

    this.group = info.group
  }
  ```

Comment (kosachevlad) [reply lesson_20_1/students.js:144]: and change this.group.studentsList to this.group everywhere below the code

File: lesson_20_1/students.js [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+  surname: "Best",
+  age: 40,
+});
+
+console.log(teacher.group.studentsList);
+console.log(teacher.getListOfNamesByAverageMark());
+console.log(teacher.getStudentByName("Marta"));
+teacher.removeStudentByName("Marta");
+console.log(teacher.getListOfNamesByAverageMark());
+console.log(teacher.updateStudentByName(eva, "Marta"));
```
Comment (kosachevlad) [inline lesson_20_1/students.js:205]: Actually not bad.

File: lesson_20_1/students.js [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+  findFakeStudent() {
+    console.log("one of possible solution is: this['#name']");
+  }
+}
+
+const teacher = new Teacher({
+  name: "Leo",
+  surname: "Best",
+  age: 40,
+});
```
Comment (kosachevlad) [inline lesson_20_1/students.js:199]: wrong instance creating...

**Review by kosachevlad — CHANGES_REQUESTED** [review]

File: lesson_20_1/students.js (hunk 1 above)
Comment (Oleksii1985) [reply lesson_20_1/students.js:18]: Done

File: lesson_20_1/students.js (hunk 2 above)
Comment (Oleksii1985) [reply lesson_20_1/students.js]: already fixed

File: lesson_20_1/students.js (hunk 3 above)
Comment (Oleksii1985) [reply lesson_20_1/students.js]: done

File: lesson_20_1/students.js (hunk 4 above)
Comment (Oleksii1985) [reply lesson_20_1/students.js]: done

File: lesson_20_1/students.js (hunk 5 above)
Comment (Oleksii1985) [reply lesson_20_1/students.js]: done

File: lesson_20_1/students.js (hunk 6 above)
Comment (Oleksii1985) [reply lesson_20_1/students.js]: ok

File: lesson_20_1/students.js (hunk 7 above)
Comment (Oleksii1985) [reply lesson_20_1/students.js]: ok

File: lesson_20_1/students.js (hunk 8 above)
Comment (Oleksii1985) [reply lesson_20_1/students.js]: done

File: lesson_20_1/students.js (hunk 9 above)
Comment (Oleksii1985) [reply lesson_20_1/students.js:144]: done

File: lesson_20_1/students.js (hunk 10 above)
Comment (Oleksii1985) [reply lesson_20_1/students.js:205]: ok

File: lesson_20_1/students.js (hunk 11 above)
Comment (Oleksii1985) [reply lesson_20_1/students.js:199]: changed

File: lesson_20_1/students.js [hunk 12]
```diff
@@ -0,0 +1,207 @@
+class Human {
+  constructor(info) {
+    this.info = {
```
Comment (kosachevlad) [inline lesson_20_1/students.js:3]: why do you need info?

File: lesson_20_1/students.js [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+      name: info.name,
+      surname: info.surname,
+      age: info.age,
+    };
+  }
+
+  getFullName() {
+    return `${this.info.name} ${this.info.surname}`;
```
Comment (kosachevlad) [inline lesson_20_1/students.js:11]: return `${this.name} ${this.surname}`;

File: lesson_20_1/students.js [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+  }
+
+  setFullName(fullName) {
+    const fullNameArray = fullName.split(" ");
+    this.info.name = fullNameArray[0];
+    this.info.surname = fullNameArray[1];
```
Comment (kosachevlad) [inline lesson_20_1/students.js:17]: this.name = fullNameArray[0];
    this.surname = fullNameArray[1];

File: lesson_20_1/students.js [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+console.log(alex.getFullName());
+alex.setFullName("Jack Daniels");
+console.log(alex.getFullName());
+
+class Student extends Human {
+  constructor(info) {
+    super(info);
+
+    this.marks = info.marks;
+  }
```
Comment (kosachevlad) [inline lesson_20_1/students.js:37]:
```javascript
 constructor({name, surname, age, marks}) {
   super({name, surname, age});
   this.marks = marks;
   }
   ```

File: lesson_20_1/students.js [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+  findFakeStudent() {
+    const fakeStudent = this.group.filter((object) => {
+      if (object instanceof FakeStudent) {
+        return object;
+      }
+    });
+    return `Fake student full name: ${fakeStudent[0].getFullName()} and real marks is: ${
+      fakeStudent[0].marks
+    }`;
+  }
```
Comment (kosachevlad) [inline lesson_20_1/students.js:191]: Well if you wanna find just one student its better to use array.find

**Review by kosachevlad — APPROVED** [review]

**Pull request merged**